### PR TITLE
Improve HF test go - Part 1

### PR DESCRIPTION
**Pull request by glyh**
This PR contains various improvement on HF-test go, and commits that are critical for https://github.com/MinaProtocol/mina/pull/17919 to be functional are all here.

### CODE REVIEW

**Comment by glyh** [comment]
!ci-nightly-me

**Comment by glyh** [comment]
^ Canceling all jobs except the hardfork test. 
https://buildkite.com/organizations/o-1-labs-2/pipelines/mina-end-to-end-nightlies/builds/3939/jobs/019a1f54-9660-4f60-bc63-368b4ad86405/log

**Comment by glyh** [comment]
!ci-build-me

File: src/app/hardfork_test/src/internal/hardfork/validation.go [hunk 1]
```diff
@@ -19,8 +19,7 @@ type BlockAnalysisResult struct {
 // ValidateSlotOccupancy checks if block occupancy is above 50%
 func (t *HardforkTest) ValidateSlotOccupancy(startingHeight, blockHeight int) error {
 	if 2*blockHeight < t.Config.BestChainQueryFrom {
-		t.Logger.Error("Assertion failed: slot occupancy is below 50%%")
```
Comment (glyh) [inline src/app/hardfork_test/src/internal/hardfork/validation.go:22]: with this log line the error would be reported twice.

File: src/app/hardfork_test/src/internal/hardfork/test.go [hunk 2]
```diff
@@ -119,7 +119,13 @@ func (t *HardforkTest) Run() error {
...
 	// Phase 1: Run and validate main network
 	t.Logger.Info("Phase 1: Running main network...")
-	forkConfigBytes, analysis, err := t.RunMainNetworkPhase(forkConfigPath, mainGenesisTs)
+	forkConfigBytes, analysis, err := t.RunMainNetworkPhase(mainGenesisTs)
+	if err != nil {
+		return err
+	}
+
+	// Write fork config to file
+	err = os.WriteFile(forkConfigPath, forkConfigBytes, 0644)
```
Comment (glyh) [inline src/app/hardfork_test/src/internal/hardfork/test.go]: TODO: write fork config after validation. could be done in follow-ups.

File: scripts/hardfork/build-and-test.sh [hunk 3]
```diff
@@ -52,9 +52,13 @@ fi
...
 SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
 
+# NOTE: Setting timezone, so if this is running in CI under NixOS, we won't 
+# panic because `/etc/localtime` being non-existent. For more, see:
+# - https://github.com/janestreet/core/blob/5a0b30cbfeb75dfc6f8d70dcceb499b250133bac/core/src/timezone_intf.ml#L27
+export TZ="UTC"
+
 if [[ $# -gt 0 ]]; then
   # Branch is specified, this is a CI run
-  ln -sf /usr/share/zoneinfo/UTC /etc/localtime
```
Comment (glyh) [inline scripts/hardfork/build-and-test.sh:57]: This timezone linking is not working in the first place. Previously localnet simply have everything thing up in background and exit code ignored. The rewrite in Go by GPT surfaced that issue

Comment (glyh) [reply scripts/hardfork/build-and-test.sh:57]: By which I mean the symlink `/usr/share` doesn't exist at all.

**Comment by glyh** [comment]
!ci-nightly-me

**Comment by glyh** [comment]
!ci-nightly-me

**Comment by glyh** [comment]
!ci-nighlty-me

**Comment by glyh** [comment]
!ci-nightly-me

**Comment by glyh** [comment]
!ci-nightly-me

File: scripts/hardfork/build-and-test.sh (hunk 3 above)
Comment (cjjdespres) [reply scripts/hardfork/build-and-test.sh:57]: Have you tried omitting the link entirely? The `core_time_intf.ml` docs in janestreet `core` say that

```
    + If TZ is not set, the system will try to read the file located at /etc/localtime,
    which must be a zic timezone database (and which is often just a symlink into
    /usr/share/zoneinfo/posix).
    + If /etc/localtime cannot be found, then the system is assumed to be in GMT.
```
and `man timezone` says

```
       If /etc/localtime is missing, the default "UTC" timezone is used.

       The timezone may be overridden for individual programs by using the $TZ environment variable. See environ(7).
```

I'm wondering if the broken symlink is confusing the timezone detection mechanism.

Comment (cjjdespres) [reply scripts/hardfork/build-and-test.sh:57]: Or could we just set the environment variable `TZ=UTC` or `TZ=Etc/UTC` for the test?

Comment (cjjdespres) [reply scripts/hardfork/build-and-test.sh:57]: If this is being run on nixos then I think the zone info should be kept at `/etc/zoneinfo`. Maybe setting `TZDIR` to that would be helpful?

Comment (cjjdespres) [reply scripts/hardfork/build-and-test.sh:57]: I'd prefer not scanning through the nix store to find `share/zoneinfo` if it can be avoided.

Comment (glyh) [reply scripts/hardfork/build-and-test.sh:57]: [Setting TZ=UTC](https://github.com/MinaProtocol/mina/commit/a51547e003793bac68158b42553fe7d17d6a07c2) doesn't work, this seems to be a bug in janestreet's core. Here's the [log](https://buildkite.com/organizations/o-1-labs-2/pipelines/mina-end-to-end-nightlies/builds/3939/jobs/019a1f54-9660-4f60-bc63-368b4ad86405/log) for that run.

I'll check if TZ=Etc/UTC works.

Comment (glyh) [reply scripts/hardfork/build-and-test.sh:57]: > If this is being run on nixos then I think the zone info should be kept at /etc/zoneinfo. Maybe setting TZDIR to that would be helpful?

We're using `gcr.io/o1labs-192920/nix-unstable:1.0.0`, if you check on it:
```bash
$ ls /etc
group  hostname  hosts	mtab  nix  passwd  resolv.conf	shadow	ssl
```

Comment (glyh) [reply scripts/hardfork/build-and-test.sh:57]: IMHO we should just abandon nix, the fact that it use so many non-standard setup enforce us to have to deal with these corner cases.
